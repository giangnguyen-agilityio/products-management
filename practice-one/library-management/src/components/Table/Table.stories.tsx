import { type StoryFn, type Meta } from "@storybook/react";
import Table from "./index";
import React from "react";

export default {
  title: "Example/Table",
  component: Table,
  argTypes: {
    onToggleCompletion: { action: "Toggle completion button clicked" },
    onEdit: { action: "Edit button clicked" },
    onDelete: { action: "Delete button clicked" },
  },
} as Meta<typeof Table>;

const Template: StoryFn<typeof Table> = (args) => <Table {...args} />;
export const Default = Template.bind({});
Default.args = {
  columns: [
    { field: "id", headerName: "ID" },
    { field: "bookId", headerName: "Book ID" },
    { field: "memberId", headerName: "Member ID" },
    { field: "fromDate", headerName: "From Date" },
    { field: "toDate", headerName: "To Date" },
    { field: "status", headerName: "Status" },
  ],
  data: [
    {
      id: "HR1",
      bookId: "B1",
      memberId: "M1",
      fromDate: "2023-03-21T08:52:13.000Z",
      toDate: "2023-07-24T08:52:13.000Z",
      status: "completed",
    },
    {
      id: "HR2",
      bookId: "B2",
      memberId: "M2",
      fromDate: "2023-03-21T08:52:13.000Z",
      toDate: "2023-03-24T08:52:13.000Z",
      status: "incomplete",
    },
    {
      id: "HR3",
      bookId: "B3",
      memberId: "M2",
      fromDate: "2023-03-21T08:52:13.000Z",
      toDate: "2023-03-24T08:52:13.000Z",
      status: "completed",
    },
    {
      id: "HR4",
      bookId: "B4",
      memberId: "M1",
      fromDate: "2023-03-21T08:52:13.000Z",
      toDate: "2023-03-24T08:52:13.000Z",
      status: "incomplete",
    },
  ],
};
