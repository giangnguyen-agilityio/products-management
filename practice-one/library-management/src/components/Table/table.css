.hire-request-table {
  border-collapse: separate;
  border-spacing: 0 10px;
  width: 100%;
}
