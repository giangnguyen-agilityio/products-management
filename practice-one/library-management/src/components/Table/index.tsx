import React, {memo} from 'react'
import TableHead from '@components/TableHead'
import TableRow from '@components/TableRow'
import {IHireRequest, TableColumn} from '@types'
import './table.css'

// Define the props for the Table component
interface TableProps {
  data: IHireRequest[]
  columns: TableColumn[]
  onToggleCompletion?: (id: string, memberId: string, bookId: string) => void
  onOpenModalEdit?: (id: string) => void
  onOpenModalDelete?: (id: string) => void
}

const Table: React.FC<TableProps> = props => {
  const {
    data,
    columns,
    onToggleCompletion,
    onOpenModalEdit,
    onOpenModalDelete,
  } = props
  return (
    // Render the hire request table
    <table className="hire-request-table">
      {/* Render the table header */}
      <TableHead columns={columns} />
      <tbody>
        {/* Render table rows for each data item */}
        {data.map(item => (
          <TableRow
            key={item.id}
            data={item}
            columns={columns}
            onToggleCompletion={onToggleCompletion}
            onOpenModalEdit={onOpenModalEdit}
            onOpenModalDelete={onOpenModalDelete}
          />
        ))}
      </tbody>
    </table>
  )
}

export default memo(Table)
