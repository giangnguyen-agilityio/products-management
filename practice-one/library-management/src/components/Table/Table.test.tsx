import React from 'react'
import {render} from '@testing-library/react'
import Table from '.'

const dataMock = [
  {
    id: '1',
    bookId: 'B01',
    memberId: 'M01',
    fromDate: '2023-07-20',
    toDate: '2023-07-27',
    status: 'completed',
  },
  {
    id: '2',
    bookId: 'B02',
    memberId: 'M02',
    fromDate: '2023-07-20',
    toDate: '2023-07-27',
    status: 'completed',
  },
]

const columnsMock = [
  {field: 'id', headerName: 'ID'},
  {field: 'bookId', headerName: 'BookId'},
  {field: 'memberId', headerName: 'MemberId'},
  {field: 'fromDate', headerName: 'From Date'},
  {field: 'toDate', headerName: 'To Date'},
  {field: 'status', headerName: 'Status'},
]

// Define mock functions
const mockToggleCompletion = jest.fn()
const mockOpenModalEdit = jest.fn()
const mockOpenModalDelete = jest.fn()

describe('Table component', () => {
  it('should render the table with the correct data', () => {
    const {container} = render(
      <Table
        data={dataMock}
        columns={columnsMock}
        onToggleCompletion={mockToggleCompletion}
        onOpenModalEdit={mockOpenModalEdit}
        onOpenModalDelete={mockOpenModalDelete}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render the table without the onToggleCompletion, onOpenModalEdit, and onOpenModalDelete handlers when they are not provided', () => {
    const {container} = render(<Table data={dataMock} columns={columnsMock} />)
    const tableControlsElement = container.querySelector('.table-controls')
    expect(tableControlsElement).toBeNull()
    expect(container).toMatchSnapshot()
  })
})
