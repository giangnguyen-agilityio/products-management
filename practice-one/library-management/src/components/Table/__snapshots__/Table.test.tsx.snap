// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render the table with the correct data 1`] = `
<div>
  <table
    class="hire-request-table"
  >
    <thead
      class="table-header"
    >
      <tr
        class="table-header-row"
      >
        <th
          class="table-header-cell"
        >
          ID
        </th>
        <th
          class="table-header-cell"
        >
          BookId
        </th>
        <th
          class="table-header-cell"
        >
          MemberId
        </th>
        <th
          class="table-header-cell"
        >
          From Date
        </th>
        <th
          class="table-header-cell"
        >
          To Date
        </th>
        <th
          class="table-header-cell"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="table-row "
      >
        <td
          class="table-row-cell"
        >
          1
        </td>
        <td
          class="table-row-cell"
        >
          B01
        </td>
        <td
          class="table-row-cell"
        >
          M01
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-20
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-27
        </td>
        <td
          class="table-row-cell"
        >
          completed
        </td>
        <td
          class="table-row-cell"
        >
          <div
            class="table-controls"
          >
            <button
              aria-label="Toggle completion button"
              class="btn btn-primary btn-medium toggle-completion-btn hide"
              type="button"
            >
              <svg
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0 1-1.08.02L2.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093L8.95 4.992a.252.252 0 0 1 .02-.022zm-.92 5.14.92.92a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 1 0-1.091-1.028L9.477 9.417l-.485-.486-.943 1.179z"
                />
              </svg>
            </button>
            <button
              aria-label="Edit button"
              class="btn btn-primary btn-medium edit-btn"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="20"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L10.582 16.07a4.5 4.5 0 01-1.897 1.13L6 18l.8-2.685a4.5 4.5 0 011.13-1.897l8.932-8.931zm0 0L19.5 7.125M18 14v4.75A2.25 2.25 0 0115.75 21H5.25A2.25 2.25 0 013 18.75V8.25A2.25 2.25 0 015.25 6H10"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
            <button
              aria-label="Delete button"
              class="btn btn-primary btn-medium delete-btn"
              type="button"
            >
              <svg
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.5 1h3a.5.5 0 0 1 .5.5v1H6v-1a.5.5 0 0 1 .5-.5ZM11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3A1.5 1.5 0 0 0 5 1.5v1H2.506a.58.58 0 0 0-.01 0H1.5a.5.5 0 0 0 0 1h.538l.853 10.66A2 2 0 0 0 4.885 16h6.23a2 2 0 0 0 1.994-1.84l.853-10.66h.538a.5.5 0 0 0 0-1h-.995a.59.59 0 0 0-.01 0H11Zm1.958 1-.846 10.58a1 1 0 0 1-.997.92h-6.23a1 1 0 0 1-.997-.92L3.042 3.5h9.916Zm-7.487 1a.5.5 0 0 1 .528.47l.5 8.5a.5.5 0 0 1-.998.06L5 5.03a.5.5 0 0 1 .47-.53Zm5.058 0a.5.5 0 0 1 .47.53l-.5 8.5a.5.5 0 1 1-.998-.06l.5-8.5a.5.5 0 0 1 .528-.47ZM8 4.5a.5.5 0 0 1 .5.5v8.5a.5.5 0 0 1-1 0V5a.5.5 0 0 1 .5-.5Z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
      <tr
        class="table-row "
      >
        <td
          class="table-row-cell"
        >
          2
        </td>
        <td
          class="table-row-cell"
        >
          B02
        </td>
        <td
          class="table-row-cell"
        >
          M02
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-20
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-27
        </td>
        <td
          class="table-row-cell"
        >
          completed
        </td>
        <td
          class="table-row-cell"
        >
          <div
            class="table-controls"
          >
            <button
              aria-label="Toggle completion button"
              class="btn btn-primary btn-medium toggle-completion-btn hide"
              type="button"
            >
              <svg
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0 1-1.08.02L2.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093L8.95 4.992a.252.252 0 0 1 .02-.022zm-.92 5.14.92.92a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 1 0-1.091-1.028L9.477 9.417l-.485-.486-.943 1.179z"
                />
              </svg>
            </button>
            <button
              aria-label="Edit button"
              class="btn btn-primary btn-medium edit-btn"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="none"
                height="20"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L10.582 16.07a4.5 4.5 0 01-1.897 1.13L6 18l.8-2.685a4.5 4.5 0 011.13-1.897l8.932-8.931zm0 0L19.5 7.125M18 14v4.75A2.25 2.25 0 0115.75 21H5.25A2.25 2.25 0 013 18.75V8.25A2.25 2.25 0 015.25 6H10"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </button>
            <button
              aria-label="Delete button"
              class="btn btn-primary btn-medium delete-btn"
              type="button"
            >
              <svg
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.5 1h3a.5.5 0 0 1 .5.5v1H6v-1a.5.5 0 0 1 .5-.5ZM11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3A1.5 1.5 0 0 0 5 1.5v1H2.506a.58.58 0 0 0-.01 0H1.5a.5.5 0 0 0 0 1h.538l.853 10.66A2 2 0 0 0 4.885 16h6.23a2 2 0 0 0 1.994-1.84l.853-10.66h.538a.5.5 0 0 0 0-1h-.995a.59.59 0 0 0-.01 0H11Zm1.958 1-.846 10.58a1 1 0 0 1-.997.92h-6.23a1 1 0 0 1-.997-.92L3.042 3.5h9.916Zm-7.487 1a.5.5 0 0 1 .528.47l.5 8.5a.5.5 0 0 1-.998.06L5 5.03a.5.5 0 0 1 .47-.53Zm5.058 0a.5.5 0 0 1 .47.53l-.5 8.5a.5.5 0 1 1-.998-.06l.5-8.5a.5.5 0 0 1 .528-.47ZM8 4.5a.5.5 0 0 1 .5.5v8.5a.5.5 0 0 1-1 0V5a.5.5 0 0 1 .5-.5Z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table component should render the table without the onToggleCompletion, onOpenModalEdit, and onOpenModalDelete handlers when they are not provided 1`] = `
<div>
  <table
    class="hire-request-table"
  >
    <thead
      class="table-header"
    >
      <tr
        class="table-header-row"
      >
        <th
          class="table-header-cell"
        >
          ID
        </th>
        <th
          class="table-header-cell"
        >
          BookId
        </th>
        <th
          class="table-header-cell"
        >
          MemberId
        </th>
        <th
          class="table-header-cell"
        >
          From Date
        </th>
        <th
          class="table-header-cell"
        >
          To Date
        </th>
        <th
          class="table-header-cell"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="table-row "
      >
        <td
          class="table-row-cell"
        >
          1
        </td>
        <td
          class="table-row-cell"
        >
          B01
        </td>
        <td
          class="table-row-cell"
        >
          M01
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-20
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-27
        </td>
        <td
          class="table-row-cell"
        >
          completed
        </td>
      </tr>
      <tr
        class="table-row "
      >
        <td
          class="table-row-cell"
        >
          2
        </td>
        <td
          class="table-row-cell"
        >
          B02
        </td>
        <td
          class="table-row-cell"
        >
          M02
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-20
        </td>
        <td
          class="table-row-cell"
        >
          2023-07-27
        </td>
        <td
          class="table-row-cell"
        >
          completed
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
